### PR TITLE
KAFKA-10787: Introduce an import order in Java sources

### PR DESCRIPTION
**Pull request by dongjinleekr**
Until now, Kafka has not enforced any import order policy. However, the lack of import order policy has been introduced many meaningless import order changes in the commits.

This PR adds an import order policy in `checkstyle.xml` that has been most generally accepted - that is, it organizes the imports to the following 4 groups in alphabetical order:

1. `org.apach.kafka.*`
2. `javax.*`
3. `java.*`
4. `*`

As an example, I reorganized 5 files in `org.apache.kafka.streams.state.internals` package here. Please pay attention to that all java package imports are grouped together and the imports on `org.apache.kafka` always go to the top of the file.

(This PR is WIP yet; As soon as this change is accpeted, I will update the PR with reorganizing all imports.)

### CODE REVIEW

**Comment by hachikuji** [comment]
Thanks, I think this makes sense. I also get a bit annoyed with the constant import shuffling. I'd like to hear some feedback from others though before we consider merging this. @ijuma any thoughts?

**Comment by ijuma** [comment]
I think this would be good. A couple of things:

1. This will generate a massive diff.
2. It would be good to have ways to format correctly automatically. It will be really annoying to do this manually. Ideally, an IntelliJ config and a Gradle plugin.
3. We need to discuss the desired order with a few committers.

**Comment by dongjinleekr** [comment]
@ijuma

> Ideally, an IntelliJ config and a Gradle plugin.

I also inspected this issue a little bit. In the case of IntelliJ, [checkstyle-idea plugin](https://plugins.jetbrains.com/plugin/1065-checkstyle-idea) works like a charm - I used this plugin.

In the case of Gradle, It seems like [diffplug/spotless](https://github.com/diffplug/spotless) plugin is the best option as of present. It has lots of features we need and is also actively developed and maintained.

> We need to discuss the desired order with a few committers.

Great. I will follow the discussion result.

**Comment by dongjinleekr** [comment]
Here is the update. 

- Adapted the three-group import ordering as discussed in [the mailing thread](https://lists.apache.org/thread.html/rf6f49c845a3d48efe8a91916c8fbaddb76da17742eef06798fc5b24d%40%3Cdev.kafka.apache.org%3E).
- Add documentation for the IDE plugins.
- Reorder the existing files and rebased onto the latest trunk.

cc/ @mjsax @cadonna @ableegoldman @vvcephei

File: clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java [hunk 1]
```diff
@@ -16,21 +16,21 @@
...
 
-import java.util.HashSet;
-import java.util.Set;
-
-import java.util.stream.Collectors;
 import org.apache.kafka.common.errors.AuthenticationException;
 import org.apache.kafka.common.utils.ExponentialBackoff;
 import org.apache.kafka.common.utils.LogContext;
+
 import org.slf4j.Logger;
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:23]: According to description, the import order is 

1. org.apach.kafka.*
1. javax.*
1. java.*
1. \*

Is ```org.slf4j.Logger``` in wrong order?

Comment (dongjinleekr) [reply clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:23]: @chia7712 Oh, the description I had at first on 2nd April is outdated; [After the discussion](https://lists.apache.org/thread.html/rf6f49c845a3d48efe8a91916c8fbaddb76da17742eef06798fc5b24d%40%3Cdev.kafka.apache.org%3E), we concluded that [the following three-group ordering](https://issues.apache.org/jira/browse/KAFKA-10787) would be better:

- `kafka`, `org.apache.kafka`
- `com`, `net`, `org`
- `java`, `javax`

**Comment by mjsax** [comment]
@dongjinleekr The PR shows merge conflict. Can you rebase it?

**Comment by mjsax** [comment]
About https://github.com/apache/kafka/pull/8404#issuecomment-645357204

As "spotless" can reformat code, could we use it to re-order imports? This might be the best solution.

**Comment by dongjinleekr** [comment]
@mjsax Rebased onto the latest trunk. Since this PR touches so many files, this PR will be broken every time the latest trunk updates. I will follow up on the latest trunk periodically as far as I can do until it is merged.

**Comment by dongjinleekr** [comment]
@mjsax And here are some explanation of why I did not adopt any auto-import reorganization or auto-formatting.

At first, I tried to add some configuration to diff/spotless plugin to reformat the code automatically. However, I found that to enable this diff/spotless requires to adopt any code formatter, like Google Java Format or Eclipse JDT Formatter, etc.

Since the current Kafka code uses [its own checkstyle format with 4 spaces](https://github.com/diffplug/spotless/issues/420), any one of the formatters above will reformat the whole codebase. So I concluded that it would be better to leave it unchanged for now.

If most of the community prefers to adopt a new formatting rule taking advantage of this chance, I will also adapt the auto-reformat plugin. It is totally up to the community's decision.

How do you think? Would it much better to open a discussion thread on dev mailing list?

cc/ @cadonna @ableegoldman @vvcephei @chia7712

**Comment by chia7712** [comment]
Personally, the import rule is good to avoid import conflicts even though there is no auto-optimization tool.

**Comment by dongjinleekr** [comment]
Rebased onto the latest trunk.

**Comment by dongjinleekr** [comment]
Closes this PR in preference of [this one](https://github.com/apache/kafka/pull/10428).

**Pull request closed without merging**